### ISSUE TITLE
Client does not start GPU core unless OpenCL is detected even if CUDA is detected

### ISSUE THREAD

**Issue by anand-bhat** [FoldingAtHome/fah-issues#1560]
## Your Environment
* **F@H Software version**:  7.6.13
* **Operating System**:  Linux
```
# cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.3 LTS (Bionic Beaver)
```
* **Browser**:  N/A

----
## Current Behavior
In a GPU environment where OpenCL is not present or detected but CUDA is available, the client adds a GPU slot and downloads GPU WUs but does not appear to call the GPU core (FAHCore22 in this instance).

Detection:
```
06:50:08:         GPUs: 1
06:50:08:        GPU 0: Bus:0 Slot:5 Func:0 NVIDIA:5 GP102GL [Quadro P6000]
06:50:08:CUDA Device 0: Platform:0 Device:0 Bus:0 Slot:5 Compute:6.1 Driver:11.0
06:50:08:       OpenCL: Not detected: clGetPlatformIDs() returned -1001
```

Entire FAHClient start up:
```
*********************** Log Started 2020-09-14T06:50:08Z ***********************
06:50:08:Trying to access database...
06:50:08:Successfully acquired database lock
06:50:08:Read GPUs.txt
06:50:08:Enabled folding slot 00: READY cpu:6
06:50:08:Enabled folding slot 01: READY gpu:0:GP102GL [Quadro P6000]
06:50:08:****************************** FAHClient ******************************
06:50:08:      Version: 7.6.13
06:50:08:       Author: Joseph Coffland <joseph@cauldrondevelopment.com>
06:50:08:    Copyright: 2020 foldingathome.org
06:50:08:     Homepage: https://foldingathome.org/
06:50:08:         Date: Apr 28 2020
06:50:08:         Time: 04:20:16
06:50:08:     Revision: 5a652817f46116b6e135503af97f18e094414e3b
06:50:08:       Branch: master
06:50:08:     Compiler: GNU 8.3.0
06:50:08:      Options: -std=c++11 -ffunction-sections -fdata-sections -O3 -funroll-loops
06:50:08:               -fno-pie
06:50:08:     Platform: linux2 4.19.0-5-amd64
06:50:08:         Bits: 64
06:50:08:         Mode: Release
06:50:08:******************************** CBang ********************************
06:50:08:         Date: Apr 25 2020
06:50:08:         Time: 00:07:53
06:50:08:     Revision: ea081a3b3b0f4a37c4d0440b4f1bc184197c7797
06:50:08:       Branch: master
06:50:08:     Compiler: GNU 8.3.0
06:50:08:      Options: -std=c++11 -ffunction-sections -fdata-sections -O3 -funroll-loops
06:50:08:               -fno-pie -fPIC
06:50:08:     Platform: linux2 4.19.0-5-amd64
06:50:08:         Bits: 64
06:50:08:         Mode: Release
06:50:08:******************************* System ********************************
06:50:08:          CPU: Intel(R) Xeon(R) CPU E5-2623 v4 @ 2.60GHz
06:50:08:       CPU ID: GenuineIntel Family 6 Model 79 Stepping 1
06:50:08:         CPUs: 8
06:50:08:       Memory: 29.44GiB
06:50:08:  Free Memory: 22.87GiB
06:50:08:      Threads: POSIX_THREADS
06:50:08:   OS Version: 4.15
06:50:08:  Has Battery: false
06:50:08:   On Battery: false
06:50:08:   UTC Offset: 0
06:50:08:          PID: 978
06:50:08:          CWD: /storage
06:50:08:           OS: Linux 4.15.0-72-generic x86_64
06:50:08:      OS Arch: AMD64
06:50:08:         GPUs: 1
06:50:08:        GPU 0: Bus:0 Slot:5 Func:0 NVIDIA:5 GP102GL [Quadro P6000]
06:50:08:CUDA Device 0: Platform:0 Device:0 Bus:0 Slot:5 Compute:6.1 Driver:11.0
06:50:08:       OpenCL: Not detected: clGetPlatformIDs() returned -1001
06:50:08:******************************* libFAH ********************************
06:50:08:         Date: Apr 15 2020
06:50:08:         Time: 21:43:24
06:50:08:     Revision: 216968bc7025029c841ed6e36e81a03a316890d3
06:50:08:       Branch: master
06:50:08:     Compiler: GNU 8.3.0
06:50:08:      Options: -std=c++11 -ffunction-sections -fdata-sections -O3 -funroll-loops
06:50:08:               -fno-pie
06:50:08:     Platform: linux2 4.19.0-5-amd64
06:50:08:         Bits: 64
06:50:08:         Mode: Release
06:50:08:***********************************************************************
06:50:08:<config>
06:50:08:  <!-- Folding Slots -->
06:50:08:  <slot id='0' type='CPU'/>
06:50:08:  <slot id='1' type='GPU'/>
06:50:08:</config>
```
GPU WU downloaded:
```
06:50:09:WU01:FS01:Requesting new work unit for slot 01: READY gpu:0:GP102GL [Quadro P6000] from 129.213.157.105
06:50:18:WU01:FS01:Received Unit: id:01 state:DOWNLOAD error:NO_ERROR project:14904 run:346 clone:2 gen:12 core:0x22 unit:0x0000001181d59d695f4ec9df64a9146c
06:50:18:WU01:FS01:Starting
06:50:18:ERROR:WU01:FS01:Failed to start core: OpenCL device matching slot 1 not found, make sure the OpenCL driver is installed or try setting 'opencl-index' manually
````

After 5 retries, the WU is returned as faulty:
```
06:59:47:ERROR:WU01:FS01:Failed to start core: OpenCL device matching slot 1 not found, make sure the OpenCL driver is installed or try setting 'opencl-index' manually
06:59:47:WU01:FS01:Sending unit results: id:01 state:SEND error:FAILED project:14904 run:346 clone:2 gen:12 core:0x22 unit:0x0000001181d59d695f4ec9df64a9146c
```
![image](https://user-images.githubusercontent.com/7332442/93058323-95705500-f6b2-11ea-96a9-3ce083b4b738.png)
I suspect this repeats until 10 WUs are returned faulty, at which point the client stops downloading new WUs until the client is restarted.

When the `opencl-index` is manually specified as `0` in `config.xml`, the client calls FAHCore22 which then runs on the CUDA platform.
```
07:41:15:WU02:FS01:0x22:Project: 14904 (Run 329, Clone 3, Gen 14)
07:41:15:WU02:FS01:0x22:Unit: 0x0000000f81d59d695f4ec9e0b7745097
07:41:15:WU02:FS01:0x22:Digital signatures verified
07:41:15:WU02:FS01:0x22:Folding@home GPU Core22 Folding@home Core
07:41:15:WU02:FS01:0x22:Version 0.0.12
07:41:15:WU02:FS01:0x22:  Checkpoint write interval: 100000 steps (5%) [20 total]
07:41:15:WU02:FS01:0x22:  JSON viewer frame write interval: 20000 steps (1%) [100 total]
07:41:15:WU02:FS01:0x22:  XTC frame write interval: 50000 steps (2.5%) [40 total]
07:41:15:WU02:FS01:0x22:  Global context and integrator variables write interval: disabled
07:41:15:WU02:FS01:0x22:There are 3 platforms available.
07:41:15:WU02:FS01:0x22:Platform 0: Reference
07:41:15:WU02:FS01:0x22:Platform 1: CPU
07:41:15:WU02:FS01:0x22:Platform 2: CUDA
07:41:15:WU02:FS01:0x22:  cuda-device 0 specified
07:41:15:WU02:FS01:0x22:opencl-device was set but OpenCL platform could not be found.
07:41:19:WU02:FS01:0x22:Attempting to create CUDA context:
07:41:19:WU02:FS01:0x22:  Configuring platform CUDA
07:41:24:WU02:FS01:0x22:  Using CUDA and gpu 0
```
----
## Expected Behavior
Do not prevent core from starting when CUDA is present and OpenCL isn't detected.

----

## Possible Solution (Optional)
With v0.0.12 of FAHCore22 now supporting CUDA, perhaps the client does not need to mandate the need for opencl support if CUDA is detected.

----
## Steps To Reproduce
1. Get a system with a GPU where OpenCL isn't available.
2. Run client.

----
## Context
There could be cases where CUDA is detected but cannot be used (e.g., Fermi devices). In such cases, I'd expect the failure to be caught at the core (no OpenCL, no usable CUDA).

----

**Comment by jcoffland** [comment]
This will change in the next client release.  It will assume OpenCL is supported if CUDA is found.

**Comment by jcoffland** [comment]
One problem, if the client cannot detect OpenCL, it cannot tell the core which OpenCL platform and device the GPU belongs to.  It give the CUDA device but then the core must be smart enough to use that info.

**Comment by anand-bhat** [comment]
Ok, I'll retain the environment with CUDA and no OpenCL and test this with the next client release to see if the core handles this

**Comment by shorttack** [comment]
Closing for now. Should be fixed in 7.18 or higher beta. Ask to reopen if this is still a problem.
